### PR TITLE
Use OpenMP to create threads in perftest

### PR DESCRIPTION
**Pull request by abouteiller**
Using Open MP to create and synchronize threads in the ucx_perftest multithreaded mode. 

It is a lot simpler and less intrusive to add the synchronization to insure non-threaded access to the rte (among other things).

### CODE REVIEW

**Comment by shamisp** [comment]
I like the idea of testing/benchmarking  with OpenMP runtime

**Comment by MattBBaker** [comment]
So on the ORNL test this fails with a bunch of link errors,
```
  CCLD   ucx_perftest
./.libs/libucxperf.a(libucxperf_la-libperf.o): In function `ucx_perf_thread_spawn.omp_fn.0':
<https://pod9.ccs.ornl.gov/job/ornl_pod9/ws/contrib/../src/tools/perf/libperf.c>:1202: undefined reference to `omp_get_thread_num'
...
```
Is this related to #959 ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2350/ for details.

**Comment by MattBBaker** [comment]
bot:ornl:retest

**Comment by abouteiller** [comment]
@MattBBaker This error seems to be a missing link flag rather than a CPPFLAG.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2417/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1761/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
This also needs to be in the Makefile.am in test/gtest/Makefile.am

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2486/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1834/ for details (Mellanox internal link).

**Comment by abouteiller** [comment]
@MattBBaker could you please email me the full output of the ORNL buildbot.

**Comment by MattBBaker** [comment]
@abouteiller Sure, what email address should I sent it to?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1848/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2538/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1891/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
I poked this more and I discovered what is going on. The pod9 system has libtool 2.2 and Mellanox has libtool 2.4. Apparently libtool 2.4 will kick up the -fopenmp flag up to the parent linker line, so you don't have to add -fopenmp yourself, while libtool 2.2 will not, so you would have to add the -fopenmp flag yourself. So it seems the question is if we want to support the six year old 2.2.6b version of libtool as seen in RHEL6.

**Comment by shamisp** [comment]
IMHO, let's go with libtool 2.4. In this case we have to add some autogen
check that will notifies user about outdated libtool. @yosefe ?

**Comment by yosefe** [comment]
does it mean UCX will not compile on RH 6?

**Comment by shamisp** [comment]
@MattB AFAIK buildbot does not use default tools, but whatever I configure
it to use.

**Comment by shamisp** [comment]
I just checked. Libtool 2.2.6 comes with RH6. What is the problem if we
explicitly add this argument ?

**Comment by MattBBaker** [comment]
@yosefe libucx and libperf will build, but unless -fopenmp is added manually to the link line of a program linking against libperf than it won't link right. The libucx libraries should be unaffected. I'm inclined to disable openmp on pod9 and say that openmp support depends on more recent tool chains.

**Comment by MattBBaker** [comment]
@shamisp Adding it explicitly in the gtest will work, but last week discussed it and tended to wonder why this didn't work when you shouldn't need to pass -fopenmp to gtest. 

The custom autotools also uses libtool 2.2.6b. I can update that to the latest 2.4, or change automake to link -fopenmp manually in gtest, or disable openmp on pod9. I'm inclined to do the latter.

**Comment by shamisp** [comment]
@MattBBaker - a.k.a disalbe openmp when libtool < 2.4.x used ?

File: src/tools/perf/uct_tests.cc [hunk 1]
```diff
@@ -238,15 +240,17 @@ class uct_perf_test_runner {
...
+#pragma omp barrier
 
         my_index = rte_call(&m_perf, group_index);
 
         ucx_perf_test_start_clock(&m_perf);
 
         buffer      = m_perf.send_buffer;
         length      = m_perf.params.message_size;
-        remote_addr = m_perf.uct.peers[1 - my_index].remote_addr;
+        remote_addr = m_perf.uct.peers[1 - my_index].remote_addr + m_perf.offset;
```
Comment (yosefe) [inline src/tools/perf/uct_tests.cc]: why do we need to add 'offset'? can't we just change remote_addr?

File: src/tools/perf/libperf.c [hunk 2]
```diff
@@ -105,7 +105,7 @@ static ucs_status_t uct_perf_test_alloc_mem(ucx_perf_context_t *perf,
     ucs_assert(perf->uct.send_mem.md == perf->uct.md);
     perf->send_buffer = perf->uct.send_mem.address;
 
-    status = uct_iface_mem_alloc(perf->uct.iface, 
+    status = uct_iface_mem_alloc(perf->uct.iface,
```
Comment (yosefe) [inline src/tools/perf/libperf.c]: pls don't do additional style fixes as part of this PR

File: src/tools/perf/libperf.c [hunk 3]
```diff
@@ -1115,53 +1117,55 @@ ucs_status_t ucx_perf_run(ucx_perf_params_t *params, ucx_perf_result_t *result)
 
 
 /* multiple threads sharing the same worker/iface */
+#include <omp.h>
```
Comment (yosefe) [inline src/tools/perf/libperf.c]: what happens if omp.h is not present?
why is it included in the middle of the file?

File: src/tools/perf/libperf.c [hunk 4]
```diff
@@ -1115,53 +1117,55 @@ ucs_status_t ucx_perf_run(ucx_perf_params_t *params, ucx_perf_result_t *result)
...
 } ucx_perf_thread_context_t;
 
+
 static void* ucx_perf_thread_run_test(void* arg) {
     ucx_perf_thread_context_t* tctx = (ucx_perf_thread_context_t*) arg;
     ucx_perf_params_t* params = &tctx->params;
     ucx_perf_result_t* result = &tctx->result;
     ucx_perf_context_t* perf = &tctx->perf;
     ucs_status_t* statuses = tctx->statuses;
-    pthread_barrier_t* tbarrier = tctx->tbarrier;
```
Comment (yosefe) [inline src/tools/perf/libperf.c]: do we remove pthread support?

**Comment by MattBBaker** [comment]
@shamisp That is the most robust solution.

**Comment by yosefe** [comment]
can we add -fopenmp manually in case of old libtool?

**Comment by shamisp** [comment]
Manually where ? Do you want to manage separate spec files..

**Comment by shamisp** [comment]
IMHO, if libtool is old , disable openmp. Otherwise enable it.

**Comment by MattBBaker** [comment]
We can add -fopenmp to the gtest link line, but that won't work if any other program wants to link against libperf and wasn't aware that -fopenmp was needed.

**Comment by abouteiller** [comment]
per discussion in the call, I have implemented the solution that add OPENMP_CFLAGS to gtest link flags.
Rationale: libucxperf is a static convenience library, so the lack of dependency propagation when using older libtools is not expected to spread issues elsewhere.

File: src/tools/perf/Makefile.am [hunk 5]
```diff
@@ -46,7 +50,10 @@ libucxperf_la_LIBADD = \
 
 ucx_perftest_SOURCES  = perftest.c
 ucx_perftest_LDFLAGS    = \
-     $(RTE_LDFLAGS)
+     $(RTE_LDFLAGS) \
+     $(OPENMP_CFLAGS)
```
Comment (abouteiller) [inline src/tools/perf/Makefile.am]: Nop, adding OPENMP_CFLAGS to LDFLAGS is the autoconf blessed solution. 
https://www.gnu.org/software/autoconf/manual/autoconf-2.67/html_node/Generic-Compiler-Characteristics.html

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2549/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1901/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2554/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1905/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
```
20:18:13   CXX      libucxperf_la-ucp_tests.lo
20:18:13   CC       perftest.lo
20:18:13 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/tools/perf/perftest.c: In function ‘sock_rte_barrier’:
20:18:13 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/tools/perf/perftest.c:601:0: error: ignoring #pragma omp master [-Werror=unknown-pragmas]
20:18:13  #pragma omp master
20:18:13  ^
20:18:13 /var/lib/jenkins/jobs/gh-ucx-pr/workspace/contrib/../src/tools/perf/perftest.c:615:0: error: ignoring #pragma omp barrier [-Werror=unknown-pragmas]
20:18:13  #pragma omp barrier
20:18:13  ^
20:18:13 cc1: all warnings being treated as errors
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2555/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1906/ for details (Mellanox internal link).

**Comment by shamisp** [comment]
@yosefe - can we merge ?

**Comment by yosefe** [comment]
@abouteiller code looks good, please squash the commits and fix commit message according to code stlye

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2577/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1933/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1950/ for details (Mellanox internal link).

**Comment by abouteiller** [comment]
I can't see the buildbot error. Is this something I should address or is this good to go?

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/9/ for details.

**Comment by yosefe** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/2629/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/27/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/1981/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:bgateNEW:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/72/ for details.

**Pull request merged**